pennylane/ops/qubit/matrix_ops.py

```diff
@@ -211,25 +226,84 @@ def __init__(
+        """Canonical matrix representation of the ControlledQubitUnitary operator.
+
+        Args:
+            U (tensor_like): unitary matrix
+            control_wires (Iterable): the control wire(s)
+            wires (Iterable): the wire(s) the unitary acts on
+            control_values (str): a string of bits representing the state of the control
+                qubits to control on (default is the all 1s state)
+
+        Returns:
+            tensor_like: canonical matrix
+
+        **Example**
+
+        >>> U = np.array([[ 0.94877869,  0.31594146], [-0.31594146,  0.94877869]])
+        >>> qml.ControlledQubitUnitary.compute_matrix(U, control_wires=[1], wires=[0], control_values="1")
+        [[ 1.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j]
+         [ 0.        +0.j  1.        +0.j  0.        +0.j  0.        +0.j]
+         [ 0.        +0.j  0.        +0.j  0.94877869+0.j  0.31594146+0.j]
+         [ 0.        +0.j  0.        +0.j -0.31594146+0.j  0.94877869+0.j]]
+        """
+        # this method reproduces the logic contained in the init function and the matrix() method
+
+        total_wires = qml.wires.Wires(control_wires) + qml.wires.Wires(wires)
+        # if control values unspecified, we control on the all-ones string
+        if not control_values:
+            control_values = "1" * len(control_wires)
+
+        control_int = int(control_values, 2)
+        padding_left = control_int * len(U)
+        padding_right = 2 ** len(total_wires) - len(U) - padding_left
+        interface = qml.math.get_interface(U)
+        left_pad = qml.math.cast_like(qml.math.eye(padding_left, like=interface), 1j)
+        right_pad = qml.math.cast_like(qml.math.eye(padding_right, like=interface), 1j)
+        return qml.math.block_diag([left_pad, U, right_pad])
+
+    def matrix(self, wire_order=None):
+        """Matrix representation of this operator.
+
+        Returns:
+            array: matrix representation
+
+        **Example**
+
+        >>> U = np.array([[ 0.94877869,  0.31594146], [-0.31594146,  0.94877869]])
+        >>> op = qml.ControlledQubitUnitary(U, control_wires=[1], wires=[0], control_values="1")
+        >>> op.matrix(wire_order=[0, 1])
+        [[ 1.        +0.j  0.        +0.j  0.        +0.j  0.        +0.j]
+         [ 0.        +0.j  0.94877869+0.j  0.        +0.j  0.31594146+0.j]
+         [ 0.        +0.j  0.        +0.j  1.        +0.j  0.        +0.j]
+         [ 0.        +0.j -0.31594146+0.j  0.        +0.j  0.94877869+0.j]]
+        """
+        if self._cu is None:
             interface = qml.math.get_interface(self.U)
             left_pad = qml.math.cast_like(qml.math.eye(self._padding_left, like=interface), 1j)
             right_pad = qml.math.cast_like(qml.math.eye(self._padding_right, like=interface), 1j)
-            self._CU = qml.math.block_diag([left_pad, self.U, right_pad])
+            self._cu = qml.math.block_diag([left_pad, self.U, right_pad])
```

Comment:
we can remove this by simply adding a cache decorator to the `compute_matrix` method no? (albeit, we would need a custom key to deal with the NumPy array, but doable: https://cachetools.readthedocs.io/en/stable/)

Reply:
Actually, there is a potential issue here --- what happens if `U` is mutated during the operations lifetime? the wrong matrix will be returned!

I'm not sure if this would ever happen (and tbh it shouldn't be a thing that's allowed), _but_ it could happen to operations with trainable parameters

Reply:
Hmm...that's true. I am happy to just compute this from scratch each time and solve speed by caching some time!

I wasn't sure if the original developer had some ideas here.

What do you think of this option: init computes CU, which is then fed as a param to the super method? We may not even have to overwrite the matrix method in that case because `compute_matrix` just gets CU through `params` and passes it again. Downside is that we carry around a larger object.

Reply:
Sounds good to me!